aws/resource_aws_sagemaker_training_job.go

```diff
@@ -0,0 +1,681 @@
+			"input_data_config": {
+				Type:     schema.TypeList,
+				MinItems: 1,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"data_source": {
+							Type:     schema.TypeList,
+							MaxItems: 1,
+							Required: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"s3_data_source": {
+										Type:     schema.TypeList,
+										MaxItems: 1,
+										Optional: true,
+										Elem: &schema.Resource{
+											Schema: map[string]*schema.Schema{
+												"s3_data_type": {
+													Type:     schema.TypeString,
+													Required: true,
+												},
+												"s3_uri": {
+													Type:     schema.TypeString,
+													Required: true,
+												},
+												"s3_data_distribution_type": {
+													Type:     schema.TypeString,
+													Required: true,
+												},
+											},
+										},
+									},
+								},
+							},
+						},
+						"content_type": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"compression_type": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "None",
+						},
+						"record_wrapper_type": {
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  "None",
+						},
+					},
+				},
+			},
+
+			"output_data_config": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Required: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"s3_output_path": {
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"kms_key_id": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"creation_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"last_modified_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"training_start_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"training_end_time": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"secondary_status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceAwsSagemakerTrainingJobCreate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).sagemakerconn
+
+	name := d.Get("name").(string)
+
+	createOpts := &sagemaker.CreateTrainingJobInput{
+		TrainingJobName: aws.String(name),
+		RoleArn:         aws.String(d.Get("role_arn").(string)),
+	}
+
+	var algorithmSpecification *sagemaker.AlgorithmSpecification = new(sagemaker.AlgorithmSpecification)
+	if as, ok := d.GetOk("algorithm_specification"); ok {
+		asPorperties := as.([]interface{})
+		properties := asPorperties[0].(map[string]interface{})
```

Comment:
We should check for nil and the length to prevent panics. `len(asProperties) > 0 && asProperties[0] != nil`